common/prometheus.cc

```diff
@@ -0,0 +1,216 @@
+/*
+ * This file is part of PowerDNS or weakforced.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 3 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "prometheus.hh"
+
+static std::shared_ptr<PrometheusMetrics> prom_metrics;
+
+void initPrometheusMetrics(std::shared_ptr<PrometheusMetrics> pmp)
+{
+  prom_metrics = pmp;
+}
+
+void PrometheusMetrics::addCommandMetric(const std::string& name)
+{
+  Counter* c = &(commands_family->Add({{"cmd", name}}));
+  commands_metrics.insert(std::make_pair(name, c));
+}
+
+void PrometheusMetrics::incCommandMetric(const std::string& name)
+{
+  auto i = commands_metrics.find(name);
+  if (i != commands_metrics.end()) {
+    i->second->Increment();
+  }
+}
+
+void PrometheusMetrics::addCustomMetric(const std::string& name)
+{
+  Counter* c = &(custom_family->Add({{"metric", name}}));
+  custom_metrics.insert(std::make_pair(name, c));
+}
+
+void PrometheusMetrics::incCustomMetric(const std::string& name)
+{
+  auto i = custom_metrics.find(name);
+  if (i != custom_metrics.end()) {
+    i->second->Increment();
```

Comment:
Not sure this is thread safe

Reply:
You mean the Increment? That class is thread-safe: https://jupp0r.github.io/prometheus-cpp/classprometheus_1_1Counter.html